crates/matrix-sdk/src/notification_settings/mod.rs

```diff
@@ -12,3 +26,751 @@ pub enum RoomNotificationMode {
     /// Do not receive any notifications.
     Mute,
 }
+
+/// A high-level API to manage the client owner's push notification settings.
+#[derive(Debug, Clone)]
+pub struct NotificationSettings {
+    /// The underlying HTTP client.
+    client: Client,
+    /// Owner's account push rules. They will be updated on sync.
+    rules: Arc<RwLock<Rules>>,
+    /// Event handler for push rules event
+    push_rules_event_handler: EventHandlerHandle,
+}
+
+impl Drop for NotificationSettings {
+    fn drop(&mut self) {
+        self.client.remove_event_handler(self.push_rules_event_handler.clone());
+    }
+}
+
+impl NotificationSettings {
+    /// Build a new `NotificationSettings``
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - A `Client` used to perform API calls
+    /// * `ruleset` - A `Ruleset` containing account's owner push rules
+    pub fn new(client: Client, ruleset: Ruleset) -> Self {
+        let rules = Arc::new(RwLock::new(Rules::new(ruleset)));
+
+        // Listen for PushRulesEvent
+        let rules_clone = rules.clone();
+        let push_rules_event_handler = client.add_event_handler(move |ev: PushRulesEvent| {
+            let rules = rules_clone.clone();
+            async move {
+                *rules.write().await = Rules::new(ev.content.global);
+            }
+        });
+
+        Self { client, rules, push_rules_event_handler }
+    }
+
+    /// Replace the internal ruleset
+    ///
+    /// # Arguments
+    ///
+    /// * `ruleset` - A `Ruleset` containing account's owner push rules
+    async fn set_ruleset(&self, ruleset: &Ruleset) {
+        *self.rules.write().await = Rules::new(ruleset.clone())
+    }
+
+    /// Get a new `Rules` instance to interact with the ruleset.
+    async fn rules(&self) -> Rules {
+        let rules = &*self.rules.read().await;
+        rules.clone()
+    }
```

Comment:
These two methods hide the fact that it's hidden behind a lock, and they may cause confusing issues, looking at how they're used:

1. get the ruleset
2. do something with it locally, start requests to the server
3. replace the internal ruleset

Since only operations 1 and 3 are holding onto the lock, there might be other operations holding onto the lock in the middle. Namely, a new `PushRulesEvent` may arrive and reset the ruleset (invalidating it in the meanwhile), or another independent request might be interleaved.

Depending on the answer to the above question (are our requests validated by receiving a new `PushRulesEvent`), maybe we don't even need to write to the local push rules. Otherwise, for operations that do read/request/writes, we need another solution:

- taking the write lock throughout the entire request would be a bit sad, but safe
- we could get the read-only `Rules`, then send the request, and after it's done apply a "diff" corresponding to the `Command` to the local `Rules` (or, equivalently, rerun the local `Rules` functions on the `self.rules.write().await` after the server has responded)
- I wonder if the server responds with a partial/updated list of push rules? In that case, we could consider the server's response is the one source of truth

Reply:
Ok, so as requested, I modified the following:
- I removed the 2 functions that were hiding the lock.
- I modified the flow used to update the ruleset. Now, every time we want to update the ruleset, we first build a list of commands that reflect the change, then we execute them to update the HS and finally, we update the local ruleset by applying the commands on the current ruleset. Applying a command can fail silently if the current ruleset has changed between the time we build the list of commands and the time we apply them to the ruleset.